keps/sig-node/3322-add-maxRestartTimes-to-podSpec/README.md

```diff
@@ -0,0 +1,973 @@
+like `Never`, then the Job API no longer needs to consider these logics anymore. Currently, it uses an annotation `batch.kubernetes.io/job-index-failure-count`
+to record the retires number, which looks not elegant.
+
+Pros:
+  * Reduce the maintenance cost of Job API
+
+Cons:
+  * The boundary of restartPolicy `Never` and `OnFailure` is vaguely.
+
+#### Caveats
+
+- Pod's maxRestartTimesOnFailure is somehow conflict with Job's backoffLimit, we'll respect the first one who hit the line.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- If `maxRestartTimesOnFailure` is configured in a recoverable circumstance with a small number retries.
+E.g. we set the `maxRestartTimesOnFailure=3`, and the network shakes with 4 times, the service will stuck
+in a failed state forever after 3 restarts, so end-users should be careful when configuring this field.
+However, this problem also exists in Job's backoffLimit, which is default to 6.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Add `maxRestartTimesOnFailure` to `pod.spec`:
+
+```golang
+type PodSpec struct {
+  // ...
+  RestartPolicy RestartPolicy
+  // maxRestartTimesOnFailure indicates the maximum count container can restart. It's the sum of
+  // all containers' restart times and only works when RestartPolicy is set to `OnFailure`.
+  // Default to nil which means will restart infinity.
+  // +optional
+  maxRestartTimesOnFailure *int
+}
+```
+
+- If Pod's `RestartPolicy` is `Always` or `Never`, `maxRestartTimesOnFailure` is default to nil, and will not apply.
+- If Pod's `RestartPolicy` is `OnFailure`, `maxRestartTimesOnFailure` is also default to nil, which means infinite
+restart times for backwards compatibility.
+
+In runtime, we'll check the sum of `RestartCount` of
+all containers [`Status`](https://github.com/kubernetes/kubernetes/blob/451e1fa8bcff38b87504eebd414948e505526d01/pkg/kubelet/container/runtime.go#L306-L335)
+packaged by Pod in function [`ShouldContainerBeRestarted`](https://github.com/kubernetes/kubernetes/blob/451e1fa8bcff38b87504eebd414948e505526d01/pkg/kubelet/container/helpers.go#L63-L96).
+If the value is greater than the `maxRestartTimesOnFailure`, return false indicating the Pod should not restart again.
+We'll add a new error named `ExceedMaxRestartTimes` as the reason of Pod's failure state.
```

Comment:
You should probably add a condition, so that it can be used from Job's `podFailurePolicy`

Reply:
My intension here is update the condition like below, do you mean I should add another condition type?
```
  - lastProbeTime: null
    lastTransitionTime: "2023-06-06T06:41:53Z"
    message: 
    reason: ExceedMaxRestartTimes
    status: "False"
    type: Ready
```

Reply:
However, I'm not fully convinced we will need it. It would be good to have a use case, @alculquicondor do you have some in mind?

If we want to add the condition, then `reason` is not matched by pod failure policy, so the `type` needs to be specific, for example:
```
type: MaxRestartTimesExceeded
status: True
reason: MaxRestartTimesExceeded
message: The number of container failures exceeded the specified maximal number of pod restarts
```

Reply:
It depends on what we want to do with backoffLimit.
Users might want to setup a rule: if max restarts exceeded, fail index/job.

Reply:
Will not add a new condition type at this moment then.